design/login.md

```diff
@@ -0,0 +1,102 @@
+reading the `WWW-Authenticate` header DC/OS returns in the 401 response:
+
+- `acsjwt` : the cluster only supports the `dcos-users` provider as defined above.
+- `oauthjwt` : the cluster only supports the OpenID implicit flow provider, as defined below:
+
+``` json
+{
+  "dcos-oidc-auth0": {
+    "authentication-type": "oidc-implicit-flow",
+    "client-method": "browser-prompt-authtoken",
+    "config": {
+      "start_flow_url": "/login?redirect_uri=urn:ietf:wg:oauth:2.0:oob"
+    },
+    "description": "Google, GitHub, or Microsoft"
+  }
+}
+```
+
+## Log in to a DC/OS cluster
+
+Log in to a DC/OS cluster is the action of presenting some credentials to the cluster in order to
+get an authorization token in return (also known as `ACS token`).
+
+The DC/OS CLI supports 6 login providers types (`authentication-type`):
+
+- **dcos-uid-password** : Log in using a standard DC/OS user account (username and password).
+- **dcos-uid-password-ldap** : Log in using an LDAP user account (username and password).
+- **dcos-uid-servicekey** : Log in using a DC/OS service user account (username and private key).
+- **saml-sp-initiated** : Log in using SAML 2.0.
+- **oidc-authorization-code-flow** : Log in using OpenID Connect authorization code flow.
+- **oidc-implicit-flow** : Log in using OpenID Connect implicit flow.
+
+A login happens through the `dcos auth login` command, which accepts the following flags:
+
+- **--provider** (string) : Specify the login provider ID to use.
+- **--username** (string) : Specify the username.
+- **--password** (string) : Specify password on the command line (insecure).
+- **--password-env** (string) : Specify an environment variable name that contains the password.
+- **--private-key** (string) : Specify the path to a file that contains the private key for service
+    account login.
+
+(A login also happens at the end of the `dcos cluster setup` command, which accepts all the flags
+from the `dcos auth login` command.)
+
+In order to perform a login, the CLI must select a login provider to use. This selection can occur
+in 3 different ways:
+
+- **explicitly** : the user passed the `--provider` flag with a given login provider ID.
+- **implicitly** : there is a single available login provider, or the user passed some flags
+    which are specific to a single login provider.
+- **manually** : when the login provider is not explicit nor implicit, a list is prompted and
+    the user is expected to select a login provider manually.
+
+Once the login provider is selected, its relevant credentials are read from command-line flags
+and the user is being prompted for the missing ones (if any). The login provider's `client-method`
+is then triggered.
+
+Each login provider has one of these 4 login methods associated to it (`client-method`):
+
+- **dcos-usercredential-post-receive-authtoken**, **dcos-credential-post-receive-authtoken** :
```

Comment:
@jgehrcke : one method is used in the `dcos-uid-password-ldap` provider definition while the other one is in the `dcos-uid-password` definition, apart from that is there any behavioral difference between the 2?